File: rulebook-cppcheck/rulebook_cppcheck/checkers/duplicate_blank_line_in_comment.py
from re import Pattern, compile as re_compile

from rulebook_cppcheck.checkers.rulebook_checkers import RulebookFileChecker
from rulebook_cppcheck.messages import Messages

try:
    from cppcheckdata import Token
except ImportError:
    from cppcheck.Cppcheck.addons.cppcheckdata import Token


class DuplicateBlankLineInCommentChecker(RulebookFileChecker):
    """See detail: https://hanggrian.github.io/rulebook/rules/#duplicate-blank-line-in-comment"""
    ID: str = 'duplicate-blank-line-in-comment'
    _MSG: str = 'duplicate.blank.line.in.comment'

    _EMPTY_COMMENT_REGEX: Pattern = re_compile(r'^\s*//\s*$')

    def check_file(self, token: Token, content: str) -> None:
        lines: list[str] = content.splitlines()
        for i in range(len(lines) - 1):
            # checks for violation
            if not self._EMPTY_COMMENT_REGEX.match(lines[i]) or \
                not self._EMPTY_COMMENT_REGEX.match(lines[i + 1]):
                continue
            self.report_error(token, Messages.get(self._MSG), i + 2)

File: rulebook-eslint/src/rules/duplicate-blank-line-in-comment.ts
import { Rule } from 'eslint';
import { Comment, SourceLocation } from 'estree';
import messages from '../messages.js';
import RulebookRule from './rulebook-rules.js';

/** {@link https://hanggrian.github.io/rulebook/rules/#duplicate-blank-line-in-comment|See detail} */
class DuplicateBlankLineInCommentRule extends RulebookRule {
    constructor() {
        super('duplicate-blank-line-in-comment', {
            [DuplicateBlankLineInCommentRule.MSG]:
                messages.get(DuplicateBlankLineInCommentRule.MSG),
        });
    }

    create(context: Rule.RuleContext) {
        return {
            'Program:exit'() {
                // checks for violation
                const comments: Comment[] =
                    context.sourceCode.getAllComments().filter(c => c.type === 'Line');
                for (let i = 0; i < comments.length - 1; i++) {
                    const current: Comment = comments[i];
                    const next: Comment = comments[i + 1];
                    const current_loc: SourceLocation | null | undefined = current.loc;
                    const next_loc: SourceLocation | null | undefined = next.loc;
                    if (!current_loc ||
                        !next_loc ||
                        current.value.trim() !== '' ||
                        next.value.trim() !== '') {
                        continue;
                    }
                    if (next_loc.start.line !== current_loc.end.line + 1) {
                        continue;
                    }
                    context.report({
                        node: next,
                        messageId: DuplicateBlankLineInCommentRule.MSG,
                    });
                }
            },
        };
    }

    static MSG: string = 'duplicate.blank.line.in.comment';
}

export default new DuplicateBlankLineInCommentRule();
